docs/architecture-decisions/0004-monolithic-application.md

```diff
@@ -0,0 +1,49 @@
+# 4. Monolithic application
+
+Date: 2023-07-20
+
+## Status
+
+Accepted
+
+## Context
+
+There are multiple ways to architect the application, or applications, to provide the type of functionality we are planning.
+
+### Monolithic Architecture
+
+One option is to build a monolithic Rails application that provides all of the planned functionality: an API to capture search terms and return structured data, the ability to analyze the search terms to determine what the structured data should be (i.e. is there a DOI in this term and if so we are confident it is for a specific item), as well as expert workflows to allow manual categorization of search terms.
+
+### Services Architecture
+
+Another option is to isolate each piece of functionality into separate services. Even within this approach, there are different ways to slice up the services to create a balance of additional application communication complexity with the benefit of single purpose applications.
+
+#### Workflows Application and Categorization Application
+
+This separates the functionality into two distinct applications.
+
+##### Workflow application
+
+The Workflow application handles staff User authentication and presents users with individual search terms that is must request from the Categorization Application. Users can write back to the Categorization Application via API when a category to assign a term to a Category. The Workflow application focuses on authentication, presenting data from an API, and posting data to an API. It is likely users would also want to see analytics about searchterms so the Categorization Application would need to expose that via API as well as the Workflow Application stores no data at all.
+
+##### Categorization Application
+
+The Categorization Application exposes multiple APIs and has no user interface that is not an API. The API must restrict write access to APIs that can write categorizations, but the categorization lookup API is likely unrestricted (even though it would store the incoming searchterm in most cases). Whether we require authentication for applications to be able to lookup data is out of scope of this ADR, but will need to be considered.
+
+Lookup API: accepts at minimum a searchterm and a string representing the calling application. Returns structured data with everythig we know about the provided search term.
+
+CategorizeThis API
+
+- GET: provides either a single term that needs categorization or possible a few terms. It is possible this may return objects or just raw terms depending on the needs of the Workflow Application.
+
+- POST: writes back a category for the provided searchterm. It is possible this will be an object or array depending on what the Workflow Application requires.
+
+Analytics API: likely a series of endpoints or a single endpoint with multiple features that will allow accept parameters and return appropriate analytics. The specific have not yet been determined.
+
+## Decision
+
+[TBD]
```

Comment:
My opinion should be understood as a naive starting point at the beginning of a long term effort - but _for the moment_ I prefer a monolithic application.

My thinking for this goes in the face of our stated architecture principles (which at a high level prefer decoupled services), but I think this is acceptable because I don't yet have a clear picture of where the appropriate decoupling would happen. The usual warnings about premature optimization are persuasive for me in this moment.

In the long term, I suspect that everything we're talking about in this repository should not live together; by starting with a monolithic approach, then, we're inviting a later refactoring to accommodate a split. That is preferable, however, to realizing later that we need to combine functions from separate applications, potentially as N components are refactored to be a differently-sized set of N components.

Reply:
I agree with everything you've said here, but I'd like to manage expectations about the possibility of a future refactor. The likelihood that we would commit resources to that feels low to me.

Reply:
If TACOS becomes useful, then considering a refactor to provide specific new features more efficiently will happen if the pain to add them to a monolith is high enough. For pathfinder work like this, I think simpler is often better and I am leaning towards a monolithic rails app that handles the categorization workflows as well as the algorithmic categorization. Having one app do both tasks has the added benefit of being able to have staff validate the algorithmic guesses as well.

The main downside is if we want to extract the algorithmic pieces into it's own API in the future, but that may be solvable even in a monolithic "feeling" application by moving them to their own Gems to be included in multiple places without reimplementation.

I guess in short, I don't see the value in moving staff categorization into a separate app from algorithmic detection. It either adds a much more complicated set of APIs that are likely to only ever be used between these two sets of features, or we cheat and have two apps share a single database (which is worse).

I'm still open to what features we gain by breaking up the monolith at this early stage, but if we can't say specifics I am not likely to be convinced.